src/cmdstan/command_helper.hpp

```diff
@@ -635,6 +672,158 @@ void services_log_prob_grad(const stan::model::model_base &model, bool jacobian,
+  if (!user_method->arg("sample"))
+    return 1;
+  unsigned int num_chains
+      = get_arg_val<int_argument>(parser, "method", "sample", "num_chains");
+  if (num_chains > 1)
+    validate_multi_chain_config(user_method);
+  return num_chains;
+}
+
+/**
+ * Check possible name conflicts between input and output files where both
+ * are in Stan CSV format; if conflicts detected, quit with an error message.
+ *
+ * @param parser user config
+ */
+void check_file_config(argument_parser &parser) {
+  std::string sample_file
+      = get_arg_val<string_argument>(parser, "output", "file");
+  std::string input_file;
+  auto user_method = parser.arg("method");
+  if (user_method->arg("generate_quantities")) {
+    input_file = get_arg_val<string_argument>(
+        parser, "method", "generate_quantities", "fitted_params");
+    if (input_file.empty()) {
+      throw std::invalid_argument(
+          std::string("Argument fitted_params file - found empty string, "
+                      "expecting filename."));
+      if (input_file.compare(sample_file) == 0) {
+        std::stringstream msg;
+        msg << "Filename conflict, fitted_params file " << input_file
+            << " and output file names are identical, must be different."
+            << std::endl;
+        throw std::invalid_argument(msg.str());
+      }
+    }
+  } else if (user_method->arg("laplace")) {
+    input_file
+        = get_arg_val<string_argument>(parser, "method", "laplace", "mode");
+    if (input_file.empty()) {
+      throw std::invalid_argument(
+          std::string("Argument mode file - found empty string, "
+                      "expecting filename."));
+      if (input_file.compare(sample_file) == 0) {
+        std::stringstream msg;
+        msg << "Filename conflict, parameter modes file " << input_file
+            << " and output file names are identical, must be different."
+            << std::endl;
+        throw std::invalid_argument(msg.str());
+      }
+    }
+  }
+}
+
+void make_filenames(const std::string &filename, const std::string &type,
+                    unsigned int num_chains, unsigned int id,
+                    std::vector<std::string> &names) {
+  names.reserve(num_chains);
+  std::string base;
+  std::string sfx;
+  get_basename_suffix(filename, base, sfx);
+  if (sfx.empty()) {
+    sfx = type;
+  }
+  auto name_iterator = [num_chains, id](auto i) {
+    if (num_chains == 1) {
+      return std::string("");
+    } else {
+      return std::string("_" + std::to_string(i + id));
+    }
+  };
+  for (int i = 0; i < num_chains; ++i) {
+    names.emplace_back(base + name_iterator(i) + sfx);
+  }
+}
+
+void init_callbacks(
+    argument_parser &parser,
+    std::vector<stan::callbacks::unique_stream_writer<std::ofstream>>
+        &sample_writers,
+    std::vector<stan::callbacks::unique_stream_writer<std::ofstream>>
+        &diag_csv_writers,
+    std::vector<stan::callbacks::json_writer<std::ofstream>>
+        &diag_json_writers) {
```

Comment:
Are all 3 of these vectors input/output arguments? If they come in as uninitialized arguments then I'd rather return a tuple or struct holding the different writers rather than have them be input / output arguments

Reply:
yes, they come in as uninitialized args, but creating a tuple struct doesn't make much sense here, given the calling structure for the services.  the main goal of `init_callbacks` is to move a chunk of tedious code out of the main part of the `command` function, in order to make it a little easier to see the overall logic flow.
not changing.